File: ios/Classes/TFLiteInterpreterHelper.swift
import TensorFlowLite

func vcpLog(_ message: String) {
    print("[VoiceCommandPlugin] \(message)")
}

final class TFLiteInterpreterHelper {
    private var interpreter: Interpreter
    private var currentInputShape: [Int]?

    init(modelPath: String) throws {
        interpreter = try Interpreter(modelPath: modelPath)
        // Some models (e.g. melspectrogram) have a default input shape that
        // overflows on allocateTensors -- they must be resized first.
        // Try eagerly; fall back to deferred allocation on first resizeInput.
        do {
            try interpreter.allocateTensors()
            let inputTensor = try interpreter.input(at: 0)
            currentInputShape = inputTensor.shape.dimensions
            vcpLog("TFLite model loaded: \(modelPath)")
            vcpLog("  Input shape: \(inputTensor.shape.dimensions), dtype: \(inputTensor.dataType)")
        } catch {
            currentInputShape = nil
            vcpLog("TFLite model loaded (deferred alloc): \(modelPath)")
        }
    }

    var inputShape: [Int]? { currentInputShape }

    func resizeInput(at index: Int, to shape: [Int]) throws {
        guard currentInputShape != shape else { return }
        try interpreter.resizeInput(at: index, to: Tensor.Shape(shape))
        try interpreter.allocateTensors()
        currentInputShape = shape
    }

    func copyInput(_ data: Data, at index: Int) throws {
        try interpreter.copy(data, toInputAt: index)
    }

    func invoke() throws {
        try interpreter.invoke()
    }

    func outputData(at index: Int) throws -> Data {
        return try interpreter.output(at: index).data
    }

    func outputFloats(at index: Int) throws -> [Float] {
        let data = try outputData(at: index)
        let count = data.count / MemoryLayout<Float>.size
        return data.withUnsafeBytes { ptr in
            guard let base = ptr.baseAddress else { return [Float]() }
            return Array(UnsafeBufferPointer(
                start: base.assumingMemoryBound(to: Float.self),
                count: count
            ))
        }
    }

    func outputShape(at index: Int) throws -> [Int] {
        return try interpreter.output(at: index).shape.dimensions
    }
}

File: android/src/main/kotlin/com/example/voice_command/TFLiteInterpreterHelper.kt
package com.example.voice_command

import android.util.Log
import org.tensorflow.lite.Interpreter
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

private const val TAG = "VoiceCommandPlugin"

internal fun vcpLog(message: String) {
    Log.d(TAG, message)
}

/**
 * Thin wrapper around TFLite [Interpreter] matching the iOS TFLiteInterpreterHelper API.
 * Supports resize → copy → invoke → read-output cycle for single-input / single-output models.
 */
internal class TFLiteInterpreterHelper(modelPath: String) {

    private val interpreter: Interpreter = Interpreter(File(modelPath))
    private var currentInputShape: IntArray? = null
    private var stagedInput: ByteBuffer? = null
    private var outputBuffer: ByteBuffer? = null

    val inputShape: IntArray? get() = currentInputShape

    init {
        try {
            interpreter.allocateTensors()
            currentInputShape = interpreter.getInputTensor(0).shape()
            vcpLog("TFLite model loaded: $modelPath")
            vcpLog("  Input shape: ${currentInputShape?.contentToString()}")
        } catch (e: Exception) {
            currentInputShape = null
            vcpLog("TFLite model loaded (deferred alloc): $modelPath")
        }
    }

    fun resizeInput(index: Int, shape: IntArray) {
        if (currentInputShape?.contentEquals(shape) == true) return
        interpreter.resizeInput(index, shape)
        interpreter.allocateTensors()
        currentInputShape = shape
    }

    fun copyInput(data: ByteBuffer, index: Int) {
        data.rewind()
        stagedInput = data
    }

    fun invoke() {
        val input = stagedInput ?: throw IllegalStateException("No input staged")
        input.rewind()

        val outTensor = interpreter.getOutputTensor(0)
        val outBuf = ByteBuffer.allocateDirect(outTensor.numBytes())
            .order(ByteOrder.nativeOrder())

        interpreter.run(input, outBuf)
        outBuf.rewind()
        outputBuffer = outBuf
        stagedInput = null
    }

    fun outputFloats(index: Int): FloatArray {
        val buf = outputBuffer ?: throw IllegalStateException("No output available at index $index")
        buf.rewind()
        val fb = buf.asFloatBuffer()
        val result = FloatArray(fb.remaining())
        fb.get(result)
        return result
    }

    fun close() {
        interpreter.close()
    }
}
